src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -108,6 +118,71 @@ struct FindComponentNearestNeighbors
           "ArborX::MST::find_component_nearest_neighbors",
           Kokkos::RangePolicy<ExecutionSpace>(space, n - 1, 2 * n - 1), *this);
     }
+
+    atomicComponentEdgesWorkaround(space);
+  }
+
+  template <typename ExecutionSpace>
+  void atomicComponentEdgesWorkaround(ExecutionSpace const &space)
+  {
+    int const n = _bvh.size();
+
+    auto &labels = _labels;           // avoid capture *this
+    auto &found_edges = _found_edges; // avoid capture *this
+    auto &edges = _edges;             // avoid capture *this
+
+    // The weights of the component edges have already been set to the correct
+    // value inside the operator(). Now, we need to set each component's edge
+    // source and target vertices.
+
+    Kokkos::parallel_for(
+        "ArborX::MST::compute_component_edge_sources",
+        Kokkos::RangePolicy<ExecutionSpace>(space, n - 1, 2 * n - 1),
+        KOKKOS_LAMBDA(int i) {
+          int component = labels(i);
+          auto const &edge = found_edges(i - n + 1);
+          auto &component_edge = edges(component - n + 1);
+
+          if (edge.weight == component_edge.weight)
+          {
+            auto min = KokkosExt::min(edge.source, edge.target);
+            Kokkos::atomic_min(&component_edge.source, min);
```

Comment:
Have you tried just comparing the whole edges here instead of splitting this into three kernels? Or is that exactly what makes #661 not solving the problem?
It seems the whole problem is that atomics on large types are pretty slow?

Reply:
Yes, this is what makes #661 slow. Apparently, on some problems, there are many equidistant edges.
And yes, atomics on larger types are slow, I think as a combination of a) they don't use hardware, and b) the operator < is expensive.